Implement to return sanitized object

andersnm opened the pull request:
This proposes to add an optional `converted` parameter to the compiled checker function whose `value` property will be assigned the converted/sanitized value. It provides the framework for #15, upon which #18 could be implemented.

The performance is down somewhat because of the additional checks (~4.1mil->3.9 mil validations/sec on my box). This could be mitigated by having different code if the converted parameter is provided or not, but I have not done this (yet?).

I assume Codacy will complain again, but I dont know how to run the same checks locally. :(

icebob commented:
Good job! Codacy said there are some indent issues only. No any serious problem.
How much work to create a different code for converted code and without it?

icebob commented:
@andersnm ?

andersnm commented:
@icebob I did commit some additional optimizations to restore the performance to pre-sanitizer for both objects and arrays :-). I'm not super happy about the API, but I think its the best we can do without breakage, and hopefully no bugs/undesired side effects.

icebob commented:
Thanks for your work. I've started to check the repo and make benchmark tests.
Yes, the "converted" solution a little bit strange, but we can change it in the future at v1.x.x

icebob commented:
Hi @andersnm,
I've added a [benchmark test for sanitization](https://github.com/andersnm/fastest-validator/blob/sanitize/benchmark/suites/sanitize.js), but it throw error. Do you know what is wrong with benchmark?

```
Suite: Sanitize simple object
× [ERR] validate & sanitize
ReferenceError: propertyValue is not defined
    at Validator.eval (eval at Validator.compileSchemaObject (D:\Work\tmp\fastest-validator\lib\validator.js:139:42), <anonymous>:20:18)
    at Validator.checkSchemaObjectInner (D:\Work\tmp\fastest-validator\lib\validator.js:261:47)
    at Validator.checkSchemaObject (D:\Work\tmp\fastest-validator\lib\validator.js:256:15)
    at D:\Work\tmp\fastest-validator\lib\validator.js:108:15
    at TestCase.bench.add [as fn] (D:\Work\tmp\fastest-validator\benchmark\suites\sanitize.js:42:12)
```

andersnm commented:
It's a bug, and found another bug as I wrote a test for it. :( Sorry, will have to get back to you later.

andersnm commented:
I've fixed the bugs I found and added tests locally, but lost steam upon realizing there should be far more tests for the sanitizing feature. Can try and commit what I have later this week

andersnm commented:
It has been rebased and force-pushed

icebob commented:
https://github.com/icebob/fastest-validator/pull/94
